src/Files.App/UserControls/StatusCenter/SpeedGraph.cs

```diff
@@ -0,0 +1,185 @@
+using Microsoft.UI.Composition;
+using Microsoft.UI.Xaml;
+using Microsoft.UI.Xaml.Controls;
+using Microsoft.UI.Xaml.Hosting;
+using Microsoft.UI.Xaml.Media;
+using System.Collections.Specialized;
+using System.Numerics;
+using Windows.UI;
+
+// To learn more about WinUI, the WinUI project structure,
+// and more about our project templates, see: http://aka.ms/winui-project-info.
+
+namespace Files.App.UserControls.StatusCenter
+{
+	public sealed partial class SpeedGraph : Control
+	{
+		public ObservableCollection<Vector2> Points
+		{
+			get => (ObservableCollection<Vector2>)GetValue(PointsProperty);
+			set
+			{
+				highestValue = 0;
+				SetValue(PointsProperty, value);
+			}
+		}
+		
+		public static readonly DependencyProperty PointsProperty =
+			DependencyProperty.Register(nameof(Points), typeof(ObservableCollection<Vector2>), typeof(SpeedGraph), null);
+
+		Compositor compositor;
+
+		ContainerVisual rootVisual;
+
+		CompositionSpriteShape graphShape;
+		InsetClip graphClip;
+
+		SpriteVisual line;
+
+		bool initialized;
+
+		float width;
+		float height;
+
+		float highestValue;
+		
+		public SpeedGraph()
+		{ 
+			// TODO: unhook
+			this.SizeChanged += OnSizeChanged;
+			this.Unloaded += UserControl_Unloaded;
+			this.ActualThemeChanged += UserControl_ActualThemeChanged;
+		}
+
+		private void OnSizeChanged(object sender, SizeChangedEventArgs e)
+		{
+			if (initialized)
+				return;
+
+			Init();
+
+			// added *after* first load
+			Points.CollectionChanged += PointsChanged;
+			this.Loaded += OnLoaded;
+		}
+
+		private void OnLoaded(object sender, RoutedEventArgs e)
+		{
+			if (Points.Count > 0)
+				UpdateGraph();
+
+			Points.CollectionChanged += PointsChanged;
+		}
+
+		private void UserControl_Unloaded(object sender, RoutedEventArgs e)
+		{
+			Points.CollectionChanged -= PointsChanged;
+		}
+
+		private void PointsChanged(object? sender, NotifyCollectionChangedEventArgs e)
+		{
+			UpdateGraph();
+		}
+
+		private void Init()
+		{
+			compositor = ElementCompositionPreview.GetElementVisual(this).Compositor;
+			rootVisual = compositor.CreateContainerVisual();
+			rootVisual.Size = this.ActualSize;
+			ElementCompositionPreview.SetElementChildVisual(this, rootVisual);
+
+			width = rootVisual.Size.X;
+			height = rootVisual.Size.Y;
+
+			var accentColor = (App.Current.Resources["AccentFillColorDefaultBrush"] as SolidColorBrush)!.Color;
+
+			var backgroundBrush = compositor.CreateColorBrush(accentColor with { A = 0x15 });
+
+			var graphFillBrush = compositor.CreateLinearGradientBrush();
+			graphFillBrush.StartPoint = new(0.5f, 0f);
+			graphFillBrush.EndPoint = new(0.5f, 1f);
+			graphFillBrush.ColorStops.Add(compositor.CreateColorGradientStop(0f, accentColor with { A = 0x70 }));
+			graphFillBrush.ColorStops.Add(compositor.CreateColorGradientStop(1f, accentColor with { A = 0x06 }));
+
+			var graphStrokeBrush = compositor.CreateColorBrush(accentColor);
+
+			var bgVisual = compositor.CreateSpriteVisual();
```

Comment:
You can use this as a root visual, there's no need to have a separate root visual